File: lib/core/models/facility.dart
// Healthcare facility model — represents a healthcare facility with GPS and capability metadata
class HealthcareFacility {
  final String id;
  final String name;
  final String type; // PHC / CHC / District Hospital / Clinic / Maternity Center
  final String address;
  final String distance; // e.g. "2.3 km"
  final String phone;
  final String hours;
  final bool isOpen;
  final List<String> services;
  final double? latitude;
  final double? longitude;
  final bool isEmergency24x7;
  final bool hasMaternalCare;

  const HealthcareFacility({
    required this.id,
    required this.name,
    required this.type,
    required this.address,
    required this.distance,
    required this.phone,
    required this.hours,
    required this.isOpen,
    required this.services,
    this.latitude,
    this.longitude,
    this.isEmergency24x7 = false,
    this.hasMaternalCare = false,
  });

  HealthcareFacility copyWith({
    String? id,
    String? name,
    String? type,
    String? address,
    String? distance,
    String? phone,
    String? hours,
    bool? isOpen,
    List<String>? services,
    double? latitude,
    double? longitude,
    bool? isEmergency24x7,
    bool? hasMaternalCare,
  }) {
    return HealthcareFacility(
      id: id ?? this.id,
      name: name ?? this.name,
      type: type ?? this.type,
      address: address ?? this.address,
      distance: distance ?? this.distance,
      phone: phone ?? this.phone,
      hours: hours ?? this.hours,
      isOpen: isOpen ?? this.isOpen,
      services: services ?? this.services,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      isEmergency24x7: isEmergency24x7 ?? this.isEmergency24x7,
      hasMaternalCare: hasMaternalCare ?? this.hasMaternalCare,
    );
  }

  factory HealthcareFacility.fromJson(Map<String, dynamic> json) {
    return HealthcareFacility(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      type: json['type'] as String? ?? '',
      address: json['address'] as String? ?? '',
      distance: json['distance'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      hours: json['hours'] as String? ?? '',
      isOpen: json['isOpen'] as bool? ?? false,
      services:
          (json['services'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      isEmergency24x7: json['isEmergency24x7'] as bool? ?? false,
      hasMaternalCare: json['hasMaternalCare'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'address': address,
      'distance': distance,
      'phone': phone,
      'hours': hours,
      'isOpen': isOpen,
      'services': services,
      'latitude': latitude,
      'longitude': longitude,
      'isEmergency24x7': isEmergency24x7,
      'hasMaternalCare': hasMaternalCare,
    };
  }
}

File: backend/src/models/Facility.js
// Mongoose model for Healthcare Facilities
const mongoose = require('mongoose');

const facilitySchema = new mongoose.Schema(
    {
        name: {
            type: String,
            required: true,
            trim: true,
            index: true,
        },
        type: {
            type: String,
            required: true,
            enum: ['Primary Health Centre', 'Community Health Centre', 'District Hospital', 'Sub Centre', 'Clinic', 'Hospital'],
            index: true,
        },
        address: {
            type: String,
            default: '',
        },
        district: {
            type: String,
            default: 'Satara',
            index: true,
        },
        state: {
            type: String,
            default: 'Maharashtra',
        },
        location: {
            type: {
                type: String,
                enum: ['Point'],
                default: 'Point',
            },
            coordinates: {
                type: [Number], // [longitude, latitude]
                default: [74.0, 17.68],
            },
        },
        distance: {
            type: String,
            default: '',
        },
        phone: {
            type: String,
            default: '',
        },
        hours: {
            type: String,
            default: 'Mon–Sat: 8am–4pm',
        },
        isOpen: {
            type: Boolean,
            default: true,
        },
        services: {
            type: [String],
            default: [],
        },
        emergencyCapable: {
            type: Boolean,
            default: false,
        },
        beds: {
            type: Number,
            default: 0,
        },
        isActive: {
            type: Boolean,
            default: true,
        },
    },
    {
        timestamps: true,
        collection: 'facilities',
    }
);

facilitySchema.index({ location: '2dsphere' });
facilitySchema.index({ name: 'text', address: 'text', services: 'text' });

module.exports = mongoose.model('Facility', facilitySchema);
